controllers/ocicluster_controller.go

```diff
@@ -106,9 +100,41 @@ func (r *OCIClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, nil
 	}
 
-	var clusterScope scope.ClusterScopeClient
+	regionOverride := r.Region
 
-	clients, err := r.ClientProvider.GetOrBuildClient(regionOverride)
+	var clusterScope scope.ClusterScopeClient
+	var clients scope.OCIClients
+	var clientProvider *scope.ClientProvider
+	if ociCluster.Spec.IdentityRef != nil {
+		identity, err := cloudutil.GetClusterIdentityFromRef(ctx, r.Client, ociCluster.Namespace, ociCluster.Spec.IdentityRef)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if !cloudutil.IsClusterNamespaceAllowed(ctx, r.Client, identity.Spec.AllowedNamespaces, ociCluster.Namespace) {
+			conditions.MarkFalse(ociCluster, infrastructurev1beta1.ClusterReadyCondition, infrastructurev1beta1.NamespaceNotAllowedByIdentity, clusterv1.ConditionSeverityError, "")
+			return reconcile.Result{}, errors.New("OCIClusterIdentity list of allowed namespaces doesn't include current cluster namespace")
+		}
+		clientProvider, err = cloudutil.GetOrBuildClientFromIdentity(ctx, r.Client, identity, regionOverride)
+		if err != nil {
+			logger.Error(err, "Could not create client provider for the cluster identity")
+			return reconcile.Result{}, err
+		}
+		region, err := clientProvider.GetRegion()
+		if err != nil {
+			logger.Error(err, "Could not get region from cluster identity")
+			return reconcile.Result{}, err
+		}
+		regionOverride = region
```

Comment:
Since the `clientProvider` was passed the `regionOverride` shouldn't the returned region be the same?

After re-reading the `GetOrBuildClientFromIdentity` this is going to trip me up. I follow the logic, but if we pass the `regionOverride` param from this level I assume we will override the Identity region (if set).

What about adding a method on the `scope.ClientProvider` with a `SetRegionIfNotSet` (terrible name) or something like that?

Reply:
did quite a bit of refactoring, please check if it looks ok